File: src/storage/meta.ts
// _meta.flop management — table registry, schema version history
// Format: "FLOP"(4B) | version(2B) | payload_len(4B) | JSON payload | CRC32(4B)

import { readUint16, writeUint16, readUint32, writeUint32, allocBuffer } from "../util/binary.ts";
import { crc32 } from "../util/crc32.ts";
import { META_FILE_MAGIC, type StoredMeta, type StoredTableMeta, type StoredSchema } from "../types.ts";

const encoder = new TextEncoder();
const decoder = new TextDecoder();

const META_VERSION = 1;

export function createEmptyMeta(): StoredMeta {
  return {
    version: META_VERSION,
    created: new Date().toISOString(),
    tables: {},
  };
}

export function serializeMeta(meta: StoredMeta): Uint8Array {
  const jsonStr = JSON.stringify(meta, null, 2);
  const payload = encoder.encode(jsonStr);

  // FLOP(4) + version(2) + payload_len(4) + payload + CRC32(4)
  const total = 4 + 2 + 4 + payload.byteLength + 4;
  const buf = allocBuffer(total);

  // Magic
  buf.set(META_FILE_MAGIC, 0);
  // Version
  writeUint16(buf, 4, META_VERSION);
  // Payload length
  writeUint32(buf, 6, payload.byteLength);
  // Payload
  buf.set(payload, 10);
  // CRC32 of payload
  const checksum = crc32(payload);
  writeUint32(buf, 10 + payload.byteLength, checksum);

  return buf;
}

export function deserializeMeta(buf: Uint8Array): StoredMeta {
  // Verify magic
  for (let i = 0; i < 4; i++) {
    if (buf[i] !== META_FILE_MAGIC[i]) {
      throw new Error("Invalid meta file: bad magic bytes");
    }
  }

  const version = readUint16(buf, 4);
  if (version !== META_VERSION) {
    throw new Error(`Unsupported meta version: ${version}`);
  }

  const payloadLen = readUint32(buf, 6);
  const payload = buf.subarray(10, 10 + payloadLen);

  // Verify CRC32
  const storedChecksum = readUint32(buf, 10 + payloadLen);
  const computedChecksum = crc32(payload);
  if (storedChecksum !== computedChecksum) {
    throw new Error("Meta file corrupted: CRC32 mismatch");
  }

  const jsonStr = decoder.decode(payload);
  return JSON.parse(jsonStr) as StoredMeta;
}

export async function readMetaFile(path: string): Promise<StoredMeta> {
  try {
    const data = await Deno.readFile(path);
    return deserializeMeta(data);
  } catch (e) {
    if (e instanceof Deno.errors.NotFound) {
      return createEmptyMeta();
    }
    throw e;
  }
}

export async function writeMetaFile(path: string, meta: StoredMeta): Promise<void> {
  const data = serializeMeta(meta);
  await Deno.writeFile(path, data);
}

export function getTableMeta(meta: StoredMeta, tableName: string): StoredTableMeta | undefined {
  return meta.tables[tableName];
}

export function setTableMeta(
  meta: StoredMeta,
  tableName: string,
  tableMeta: StoredTableMeta,
): void {
  meta.tables[tableName] = tableMeta;
}

export function createTableMeta(schema: StoredSchema): StoredTableMeta {
  return {
    currentSchemaVersion: 1,
    schemas: { 1: schema },
  };
}

export function addSchemaVersion(
  tableMeta: StoredTableMeta,
  schema: StoredSchema,
): number {
  const newVersion = tableMeta.currentSchemaVersion + 1;
  tableMeta.schemas[newVersion] = schema;
  tableMeta.currentSchemaVersion = newVersion;
  return newVersion;
}

File: internal/storage/meta.go
package storage

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/marcisbee/flop/internal/schema"
	"github.com/marcisbee/flop/internal/util"
)

const metaVersion = 1

// CreateEmptyMeta returns a fresh StoredMeta.
func CreateEmptyMeta() *schema.StoredMeta {
	return &schema.StoredMeta{
		Version: metaVersion,
		Created: time.Now().UTC().Format(time.RFC3339Nano),
		Tables:  make(map[string]*schema.StoredTableMeta),
	}
}

// SerializeMeta encodes StoredMeta to _meta.flop binary format.
func SerializeMeta(meta *schema.StoredMeta) ([]byte, error) {
	payload, err := json.MarshalIndent(meta, "", "  ")
	if err != nil {
		return nil, err
	}

	// FLOP(4) + version(2) + payload_len(4) + payload + CRC32(4)
	total := 4 + 2 + 4 + len(payload) + 4
	buf := make([]byte, total)

	copy(buf[0:4], schema.MetaFileMagic[:])
	binary.LittleEndian.PutUint16(buf[4:6], metaVersion)
	binary.LittleEndian.PutUint32(buf[6:10], uint32(len(payload)))
	copy(buf[10:], payload)
	checksum := util.CRC32(payload)
	binary.LittleEndian.PutUint32(buf[10+len(payload):], checksum)

	return buf, nil
}

// DeserializeMeta decodes a _meta.flop file.
func DeserializeMeta(data []byte) (*schema.StoredMeta, error) {
	if len(data) < 14 {
		return nil, fmt.Errorf("meta file too short")
	}
	// Verify magic
	for i := 0; i < 4; i++ {
		if data[i] != schema.MetaFileMagic[i] {
			return nil, fmt.Errorf("invalid meta file: bad magic bytes")
		}
	}
	version := binary.LittleEndian.Uint16(data[4:6])
	if version != metaVersion {
		return nil, fmt.Errorf("unsupported meta version: %d", version)
	}
	payloadLen := binary.LittleEndian.Uint32(data[6:10])
	if int(10+payloadLen+4) > len(data) {
		return nil, fmt.Errorf("meta file truncated")
	}
	payload := data[10 : 10+payloadLen]
	storedChecksum := binary.LittleEndian.Uint32(data[10+payloadLen:])
	computedChecksum := util.CRC32(payload)
	if storedChecksum != computedChecksum {
		return nil, fmt.Errorf("meta file corrupted: CRC32 mismatch")
	}

	var meta schema.StoredMeta
	if err := json.Unmarshal(payload, &meta); err != nil {
		return nil, fmt.Errorf("meta file JSON parse error: %w", err)
	}
	return &meta, nil
}

// ReadMetaFile reads _meta.flop from disk.
func ReadMetaFile(path string) (*schema.StoredMeta, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return CreateEmptyMeta(), nil
		}
		return nil, err
	}
	return DeserializeMeta(data)
}

// WriteMetaFile writes _meta.flop to disk.
func WriteMetaFile(path string, meta *schema.StoredMeta) error {
	data, err := SerializeMeta(meta)
	if err != nil {
		return err
	}
	return writeFileAtomic(path, data, 0o644)
}

// CreateTableMeta creates a new table meta with initial schema.
func CreateTableMeta(stored *schema.StoredSchema) *schema.StoredTableMeta {
	return &schema.StoredTableMeta{
		CurrentSchemaVersion: 1,
		Schemas:              map[int]*schema.StoredSchema{1: stored},
	}
}

// AddSchemaVersion bumps the schema version and records the new schema.
func AddSchemaVersion(tableMeta *schema.StoredTableMeta, stored *schema.StoredSchema) int {
	newVersion := tableMeta.CurrentSchemaVersion + 1
	tableMeta.Schemas[newVersion] = stored
	tableMeta.CurrentSchemaVersion = newVersion
	return newVersion
}
